### PR TITLE
Use DMA Transfer API in I2S module

### PR DESCRIPTION
**Pull request by kalkyl**
I changed the I2S module to use a DMA `Transfer` API for TX and RX operations.
I also added support for full duplex transfers and the demos were updated according to the new API.

"Controller mode" demo (TX):
https://github.com/kalkyl/nrf-hal/blob/i2s-dma/examples/i2s-controller-demo/src/main.rs

"Peripheral mode" demo (RX):
https://github.com/kalkyl/nrf-hal/blob/i2s-dma/examples/i2s-peripheral-demo/src/main.rs

### CODE REVIEW

**Comment by jonas-schievink** [comment]
Thanks!

I think `Pin` might not be the right thing to use for DMA APIs (see https://github.com/rust-embedded/embedonomicon/issues/64). The [`embedded-dma`](https://docs.rs/embedded-dma/0.1.0/embedded_dma/) crate provides traits for DMA APIs though, I wonder if we should just use those.

**Comment by kalkyl** [comment]
> Thanks!
> 
> I think `Pin` might not be the right thing to use for DMA APIs (see [rust-embedded/embedonomicon#64](https://github.com/rust-embedded/embedonomicon/issues/64)). The [`embedded-dma`](https://docs.rs/embedded-dma/0.1.0/embedded_dma/) crate provides traits for DMA APIs though, I wonder if we should just use those.

Ahh i see... I can't really figure out how to implement them for this at the moment, i guess i haven't fully understood how they're supposted to be used... Any hints?

**Comment by jonas-schievink** [comment]
I've also never used them so far, unfortunately.

**Comment by kalkyl** [comment]
I’ll do my homework ;)

**Comment by kalkyl** [comment]
Ok i have a working implementation using the `embedded_dma` traits. The only problem is  `embedded_dma` doesn't implement `Word` for signed integer types, so we'd need upstream support for that to make it usable.

**Comment by eldruin** [comment]
FYI, I just published `embedded-dma` 0.1.1

File: nrf-hal-common/src/i2s.rs [hunk 1]
```diff
@@ -341,28 +330,24 @@ impl I2S {
...
-            return Err(Error::DMABufferNotInDataMemory);
-        }
-
-        if buffer.maxcnt() > MAX_DMA_MAXCNT {
+        let (ptr, len) = unsafe { buffer.read_buffer() };
+        let maxcnt = (len / (core::mem::size_of::<u32>() / core::mem::size_of::<W>())) as u32;
+        if maxcnt > MAX_DMA_MAXCNT {
             return Err(Error::BufferTooLong);
         }
+        if (ptr as usize) < SRAM_LOWER || (ptr as usize) > SRAM_UPPER {
```
Comment (jonas-schievink) [inline nrf-hal-common/src/i2s.rs:342]: Wouldn't this have to be

```suggestion
        if (ptr as usize) < SRAM_LOWER || (ptr as usize + maxcnt) > SRAM_UPPER {
```

to check for the *whole* buffer, not just its start?

Comment (kalkyl) [reply nrf-hal-common/src/i2s.rs:342]: Ahh, i was thinking it'd always be located either "inside RAM" or not?

Comment (jonas-schievink) [reply nrf-hal-common/src/i2s.rs:342]: Hmm, yeah, that's true

**Review by jonas-schievink — APPROVED** [review]
Looks good, thanks!

**Pull request merged**